src/main/java/mate/academy/AsyncRequestProcessor.java

```diff
@@ -1,16 +1,29 @@
 package mate.academy;
 
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
 
 public class AsyncRequestProcessor {
+    private final ConcurrentHashMap<String, UserData> cache = new ConcurrentHashMap<>();
     private final Executor executor;
 
     public AsyncRequestProcessor(Executor executor) {
         this.executor = executor;
     }
 
     public CompletableFuture<UserData> processRequest(String userId) {
-        return null;
+        return cache.get(userId) != null ? CompletableFuture.completedFuture(cache.get(userId)) :
+                CompletableFuture.supplyAsync(() -> getUserDataByUserId(userId), executor)
+                        .whenCompleteAsync((u, ex) -> {
+                            if (ex != null) {
+                                throw new RuntimeException();
+                            }
+                            cache.put(u.userId(), u);
+                        });
+    }
+
+    public UserData getUserDataByUserId(String userId) {
```

Comment:
Why is this method public?

Reply:
Actually, I don't know why I used modifier public )) Changed it to private